replication/client/client.go

```diff
@@ -280,51 +292,63 @@ func (c *Replicator) Start(progressChan <-chan uint64) {
 		}
 
 		// Get a connection
-		conn, rplErr = c.connManager.GetConn()
+		pgConn, rplErr = c.connManager.GetConn(c.shutdownHandler.TerminateCtx)
 		if rplErr != nil {
 			log.Error(rplErr.Error())
 			return
 		}
 
 		// Setup a timeout and then read from replication slot on server
 		func() {
-			replicationCtx, cancelFn := context.WithTimeout(c.shutdownHandler.TerminateCtx, 5*time.Second) // TODO should this branch from our context?
+			replicationCtx, cancelFn := context.WithTimeout(c.shutdownHandler.TerminateCtx, 5*time.Second)
 			defer cancelFn()
 
-			message, rplErr = conn.WaitForReplicationMessage(replicationCtx)
+			message, rplErr = pgConn.ReceiveMessage(replicationCtx)
 		}()
 
 		// Check for error and connection status
 		if rplErr != nil {
-			// check if deadline was exceeded
-			if rplErr.Error() == "context deadline exceeded" {
-				log.Debug("deadline exceeded")
-
-				// Send a keepalive whenever deadline is exceeded. This keeps
-				// the connection alive when there are no new messages from
-				// postgres.
+			if pgconn.Timeout(rplErr) {
 				if err := c.handleProgress(true); err != nil {
 					log.Error(err)
 					return
 				}
+
 				continue
 			}
 
+			log.WithError(rplErr).Warn("handling replication error")
+
 			// TODO(#8): what about other types of errors?
-			if !conn.IsAlive() {
+			if pgConn.IsClosed() {
 				log.Warn("connection was closed")
 				continue
 			}
+
+			return
 		}
 
+		// Begin reading message
 		if message == nil {
 			log.Debug("message was nil")
 			continue
 		}
 
+		cd, ok := message.(*pgproto3.CopyData)
+		if !ok {
+			log.Errorf("Received unexpected message: %#v", message)
+			continue
```

Comment:
What happens here?

Reply:
Same as above. I haven't handled this yet but it will be part of this PR.